### ISSUE TITLE
`InitializeUserDetailsBeanManagerConfigurer` looks for `UserDetailsService` even when using OAuth

### ISSUE THREAD

**Issue by mbhave** [spring-projects/spring-security#4795]

**Comment by snicoll** [comment]
We need this in order to fix an issue for RC1. Can we please triage this one?

**Comment by jgrandja** [comment]
Yes, `InitializeUserDetailsBeanManagerConfigurer` looks for a `UserDetailsService` bean and if registered it will create a `DaoAuthenticationProvider` (composing the `UserDetailsService` bean) which will be aggregated in the **global** `AuthenticationManager`. All of this happens in `AuthenticationConfiguration`.

The fact that Boot's `AuthenticationManagerConfiguration` automatically registers a `UserDetailsService` bean of type `InMemoryUserDetailsManager`, this triggers the building of the **global** `AuthenticationManager` and ultimately the log output of the default password which is causing this issue.

The `@Lazy` solution for `InMemoryUserDetailsManager` (instead of `@Bean`) will not work because the 
**global** `AuthenticationManager` **must be built** before it's associated to each **local** `AuthenticationManager` which is associated to a `WebSecurityConfigurerAdapter` (or `FilterChain`), for example, Boot's `OAuth2WebSecurityConfigurerAdapter`.

Again to re-iterate, even if `InMemoryUserDetailsManager` was marked as `@Lazy`, it will get eagerly initialized when the **global** `AuthenticationManager` is built. And it must be built before it's associated to a **local** `AuthenticationManager` of a `WebSecurityConfigurerAdapter`.

The good news is there is a solution that is quick and easy to implement. I've tested this on the latest Boot snapshot and it works.

In `AuthenticationManagerConfiguration`, add this:

```
	@ConditionalOnMissingBean(type = "org.springframework.security.oauth2.client.registration.ClientRegistrationRepository")
	@Bean
	public InMemoryUserDetailsManager inMemoryUserDetailsManager(
```

So if `ClientRegistrationRepository` is not registered than register `InMemoryUserDetailsManager`.

What do you think?

**Comment by jgrandja** [comment]
One other thing I wanted to point out is, even if we can modify code on our end to not initialize the `UserDetailsService` bean, it will still be registered in the `ApplicationContext` and never used.

IMO, if a bean is never used than it doesn't make sense to have it registered.

**Comment by mbhave** [comment]
@jgrandja Is there a use-case where `InMemoryUserDetailsManager` needs to be aggregated in the global AuthenticationManager even when OAuth is used? If not, I think Spring Security should not initialize the `UserDetailsService` bean.

**Comment by rwinch** [comment]
There are instances where a user may want to support OAuth and form based log in. In those cases, we would need a `UserDetailsService` along with OAuth.

I'd like to emphasize Joe's last point. Let's assume that we go with my initial proposal (I realize since suggesting it that it was wrong) of using `@Lazy` and modify Spring Security to not access `InMemoryUserDetailsManager` when OAuth is being used.  It does fix the logging issue. However, this  means that there is an `InMemoryUserDetailsManager` Bean definition that is not being used. To me this is reason enough to avoid using the `@Lazy` approach. We do not want unnecessary Bean definitions in the `ApplicationContext`.

**Comment by mbhave** [comment]
> There are instances where a user may want to support OAuth and form based log in. In those cases, we would need a UserDetailsService along with OAuth.

Doesn't that mean that we can't base the creation of `UserDetailsService` bean on the presence of a `ClientRegistrationRepository` bean?

> We do not want unnecessary Bean definitions in the ApplicationContext.

Agreed.

**Comment by rwinch** [comment]
> Doesn't that mean that we can't base the creation of UserDetailsService bean on the presence of a ClientRegistrationRepository bean?

It does in the same way that the presence of an `AuthenticationManager` might be used in conjunction with an additional `UserDetailsService`. The configuration options/permuations are just too complex to guess what a user is doing.

**Comment by jgrandja** [comment]
> Is there a use-case where InMemoryUserDetailsManager needs to be aggregated in the global AuthenticationManager even when OAuth is used? If not, I think Spring Security should not initialize the UserDetailsService bean

Yes, this is a real world use-case as there are many websites that allow users to authenticate using their existing credentials at an external OAuth provider (e.g. Login with Google or Facebook) OR login with the credentials they entered during profile/account registration. In these cases, `http.formLogin()` and `http.oauth2Login()` would be configured along with a `UserDetailsService` bean that would be used when the user logs in via `http.formLogin()`.

Moreover, even if there wasn't a valid use case, I don't see a valid reason to default the `UserDetailsService` bean to `@Lazy`. If the user provides their own `UserDetailsService` bean than it would be `@Lazy` by default and will not be what the user expects or wants. But now they don't have that control to force eager initialization. No doubt a ticket would be logged down the road to reverse this effect and we would need to deal with this again.

**Comment by jgrandja** [comment]
> Doesn't that mean that we can't base the creation of UserDetailsService bean on the presence of a ClientRegistrationRepository bean?

To elaborate further from @rwinch comment...if the user has provided their own `WebSecurityConfigurerAdapter` than Boot will back-off from auto-configuration and the user will apply `http.oauth2Login()` and `http.formLogin()` and whatever else they choose to configure. However, if the user does not provide a `WebSecurityConfigurerAdapter` than auto-configuration kicks in and the solution provided earlier makes sense (at least to me) to exclude the registration of `InMemoryUserDetailsManager` when `ClientRegistrationRepository` is available.
